### ISSUE TITLE
VisibilityConfig for static elements in Form

### ISSUE THREAD

**Issue by aviriel** [Alfresco/Aikau#1047]
VisibilityConfig is not working for layout and other static elements like "alfresco/forms/ControlRow" or simple "alfresco/html/Label" on the form.
Sample:
```
{
	name: "alfresco/forms/DynamicForm",
	config: {
		....
		widgets: [{
			name: "mgs/widgets/forms/Select",
			config: {
				name: "caseType",
				label: "Case Type",
				fieldId: "PRESET_CASETYPE",
				value: "1",
				optionsConfig: {
					fixed: [{
						"label": "1",
						"value": "1"
					}, {
						"label": "2",
						"value": "2"
					}, {
						"label": "3",
						"value": "3"
					}]
				}
			}
		}, {
			name: "alfresco/forms/controls/TextBox",
			config: {
				name: "TypeCode",
				label: "Type Code",
				value: "",
				visibilityConfig: {
					initialValue: true,
					rules: [{
						"targetId": "PRESET_CASETYPE",
						"is": [
							"1"
						]
					}]
				}
			}
		}, {
			name: "alfresco/html/Label",
			config: {
				label: "Label",
				visibilityConfig: {
					initialValue: true,
					rules: [{
						"targetId": "PRESET_CASETYPE",
						"is": ["1"],
						strict: true
					}]
				}
			}
		}, {
			name: "alfresco/forms/ControlRow",
			widthPx: 600,
			config: {
				widgetMarginLeft: 5,
				widgetMarginRight: 0,
				widgets: [{
					name: "alfresco/forms/controls/DateTextBox",
					widthPx: 185,
					config: {
						label: "From",
						name: "prop_DocumentDate1",
						value: ""
					}
				}, {
					name: "alfresco/forms/controls/DateTextBox",
					widthPx: 200,
					config: {
						label: "To",
						name: "prop_DocumentDate2",
						value: ""
					}
				}],
				visibilityConfig: {
					initialValue: true,
					rules: [{
						"targetId": "PRESET_CASETYPE",
						"is": ["1"],
						strict: true
					}]
				}
			}
		}]
	}
}
```
TextBox is updated every time I change the value of Select (and all other children of BaseFormControl), but Label and ControlRow ignore changes. Label works for me if I use it not in the Form. Any ideas what to do with it?

I'm on 1.0.60, but as I can see there were no changes in this part of code since that time.

**Comment by AFaust** [comment]
The visibilityConfig property works slightly different in form controls than in non-form widgets. Only form controls can use the 'targetId' property to refer to other form controls. Any other widgets can only use the standard 'topic' attribute. For a widget like 'alfresco/html/Label' you should be able to make it work by using
```
{
   visibilityConfig : {
      initialValue : true,
      rules : [{
         topic : '_valueChangeOf_PRESET_CASETYPE',
         attribute : 'value',
         is : ['1'],
         strict : true
      }]
   }
}
```

**Comment by aviriel** [comment]
It works, thank you. I'm confused with the number of poosible attributes.
The next step will be "How to configure complex conditions for RulesEngine, like visible if (A AND B) OR (C AND D)?".

**Comment by aviriel** [comment]
Heh, no, it does not work.

```
{
    name: "alfresco/forms/DynamicForm",
    config: {
        ....
        widgets: [{
            name: "mgs/widgets/forms/Select",
            config: {
                name: "A",
                label: "A",
                fieldId: "PRESET_A",
                value: "1",
                optionsConfig: {
                    fixed: [{
                        "label": "1",
                        "value": "1"
                    }, {
                        "label": "2",
                        "value": "2"
                    }, {
                        "label": "3",
                        "value": "3"
                    }]
                }
            }
        }, {
            name: "mgs/widgets/forms/Select",
            config: {
                name: "B",
                label: "B",
                fieldId: "PRESET_B",
                value: "1",
                optionsConfig: {
                    fixed: [{
                        "label": "1",
                        "value": "1"
                    }, {
                        "label": "2",
                        "value": "2"
                    }, {
                        "label": "3",
                        "value": "3"
                    }]
                },
                visibilityConfig: {
                    initialValue: true,
                    rules: [{
                        "targetId": "PRESET_A",
                        "is": ["1", "2"],
                        strict: true
                    }]
                }
            }
        }, {
            name: "alfresco/html/Label",
            config: {
                label: "Label",
                visibilityConfig: {
                    initialValue: true,
                    rulesMethod: "ALL",
                    rules: [{
                        topic: "_valueChangeOf_PRESET_A",
                        attribute: 'value',
                        is: [ "1", "2" ],
                        strict: true
                    }, {
                        topic: "_valueChangeOf_PRESET_B",
                        attribute: 'value',
                        is: [ "1" ],
                        strict: true
                    }]
                }
            }
        }, {
            name: "alfresco/html/Label",
            config: {
                label: "Label2",
                visibilityConfig: {
                    initialValue: true,
                    rulesMethod: "ALL",
                    rules: [{
                        topic: "_valueChangeOf_PRESET_A",
                        attribute: 'value',
                        is: [ "1", "2" ],
                        strict: true
                    }, {
                        topic: "_valueChangeOf_PRESET_B",
                        attribute: 'value',
                        is: [ "2" ],
                        strict: true
                    }]
                }
            }
        }]
    }
}
```

If I update A, then B, then it works, I see either Label or Label2. If I update B, then A, then it shows all labels.

**Comment by AFaust** [comment]
(In)Visibility config on normal widgets does not support a rulesMethod attribute - this is another feature that only form controls have by inheriting from RulesEngineMixin. For normal widgets, the visibility is basically re-evaluated on every topic pubSub and only the result of the rule(s) that apply to that topic are considered. The order that you observe regarding "works" / "doesn't work" is due to your rule configuration (same config for rule A, different for rule B).

This is a short-coming of the CoreWidgetProcessing module that prompted me to hack it in our project to add state, storing previous results of individual rules and resolving conflicts between strict/non-strict rules. As this is an invasive change and I have not yet managed to find the time to set up an isolated test VM, I have not turned that into a PR yet.

**Comment by aviriel** [comment]
I see. Ok, I will try to solve it somehow. Should I leave this issue opened and rename it or just close it?

**Comment by AFaust** [comment]
Leave it open for @draperd or @MartinDoyleUK to see and respond. I only did because I was familiar with the problems...

**Comment by draperd** [comment]
Apologies for the delay in responding to this issue - the Aikau team were on vacation, we will start looking through the backlog of issues from today - thanks for your patience!

**Comment by draperd** [comment]
Can I just clarify if this should be treated as a separate issue from #1049 - or if that issue supercedes this one? I'd just like to know if we need another JIRA ticket open.

**Comment by aviriel** [comment]
It's a separate issue. Finally it relates to impossibility to add several rules to visibilityConfig for widgets like "alfresco/html/Label". If I use label on the form, then I can configure dependency on the TextBox on other fields but I cannot do it for Label:
```
{
    name: "alfresco/forms/DynamicForm",
    config: {
        ....
        widgets: [{
            name: "mgs/widgets/forms/Select",
            config: {
                name: "A",
                label: "A",
                fieldId: "PRESET_A",
                value: "1",
                optionsConfig: {
                    fixed: [{
                        "label": "1",
                        "value": "1"
                    }, {
                        "label": "2",
                        "value": "2"
                    }, {
                        "label": "3",
                        "value": "3"
                    }]
                }
            }
        }, {
            name: "mgs/widgets/forms/Select",
            config: {
                name: "B",
                label: "B",
                fieldId: "PRESET_B",
                value: "1",
                optionsConfig: {
                    fixed: [{
                        "label": "1",
                        "value": "1"
                    }, {
                        "label": "2",
                        "value": "2"
                    }, {
                        "label": "3",
                        "value": "3"
                    }]
                }
            }
        }, {
            name: "alfresco/html/Label",
            config: {
                label: "Label",
                visibilityConfig: {
                    initialValue: true,
                    rulesMethod: "ALL",
                    rules: [{
                        topic: "_valueChangeOf_PRESET_A",
                        attribute: 'value',
                        is: [ "1", "2" ],
                        strict: true
                    }, {
                        topic: "_valueChangeOf_PRESET_B",
                        attribute: 'value',
                        is: [ "1" ],
                        strict: true
                    }]
                }
            }
        }, {
            name: "alfresco/forms/controls/TextBox",
            config: {
                name: "DependentTextBox",
                label: "DependentTextBox",
                value: "",
                visibilityConfig: {
                    initialValue: true,
                    rulesMethod: "ALL",
                    rules: [{
                        "targetId": "PRESET_A",
                        "is": [
                            "1", "2"
                        ]
                    }, {
                        "targetId": "PRESET_B",
                        "is": [
                            "2"
                        ]
                    }]
                }
            }
        }]
    }
}
```

**Comment by draperd** [comment]
OK, thanks - I'll raise a separate ticket then - just catching up with all the issues, e-mails, etc after a week or so away.

**Comment by draperd** [comment]
I've raised https://issues.alfresco.com/jira/browse/AKU-974 to investigate and fix this issue.

**Comment by draperd** [comment]
This is an interesting scenario and one that I hadn't expect people to have been a requirement. Now that I've tested out the model I can see that @AFaust 's comment is absolutely correct. The issue here is that a form is expected to only contain a small set of values for evaluation (e.g. the available fields).  It should be possible to expand the behaviour of visibilityConfig of all widgets to keep track of every possible publication however I would be concerned about potential breakages of existing pages (I'm not sure about this, it's just a hunch that certain things could be impacted).

However, it might be possible and more efficient for us to provide an additional "opt-in" configuration setting that instructs the widget to track all publication changes and this would allow more complex rules to be defined as with the form controls.

I'll do some investigation into this approach. I would be interested to see the code that @AFaust  has implemented - even if not extrapolated into an individual pull request - if at all possible!

**Comment by aviriel** [comment]
Thanks for your response, David. As a workaround I decided to create a hidden field that changes the value automatically depends on 2 other fields (#1049), and then I can configure visibility config of the label as dependency of this new one field. It works, but it will be interesting to see if it's possible to do it in other way.

**Comment by draperd** [comment]
The approach I've taken so far is to mixin the RulesEngineMixin module into CoreWidgetProcessing to make the same rules engine that is used for form controls be applicable for any widget. However, as mentioned in my previous comment this will take an "opt-in" approach for backwards compatibility.

For example configuration would now look like:

```JAVASCRIPT
visibilityConfig: {
   useFormStyle: true,
   initialValue: true,
   rulesMethod: "ALL",
   rules: [
      {
         targetId: "PRESET",
         attribute: "value",
         is: ["11"],
         strict: true
      },
      {
         targetId: "PRESET_A",
         attribute: "value",
         is: [ "1" ],
         strict: true
      },
      {
         targetId: "PRESET_B",
         attribute: "value",
         is: [ "1" ],
         strict: true
     }
   ]
}
```

Note the use of the `useFormStyle` attribute in `visibilityConfig` - this switches rules evaluation to be the same as forms based. This in turn means that `targetId` is used instead of `topic` and it is not necessary to include "_valueChangeOf_" as a prefix in the topic. Using this option will make any widget behave exactly like a form control.

I've not submitted this change for review yet as I was looking to get some feedback first.... one thing that has occurred to me is that this means that this approach will **only** work within forms - so I suspect my changes don't go as far as those of @AFaust (this is because of the switch to `targetId` over topic). If this is a concern, then I could possibly look to support both `targetId` and `topic` with `useFormStyle`.

**Comment by AFaust** [comment]
Well, your changes are a bit different from mine. Using the rule engine addresses "state handling" features that we have included in our CoreWidgetProcessingFixMixin, e.g. rules remember the previous pass/fail result. We have implemented a logic that relied on "strict" instead of an additional "rulesMethod" to determine how rule results should be aggregated, but I prefer an explicit, easy-to-understand "rulesMethod" over our logic which tip-toed around and had to work with what was already defined/used.

The only thing I'd like to vote against is the "either-or" style of ```topic``` vs ```targetId```. I would argue that both form and normal widget visibility have valid use cases where both styles would need to be combined into a rule set. I.e. a form control may very well have to react to an external button publish to show/hide. Why not support both ```topic``` and ```targetId``` in CoreWidgetProcessing based on presence of attribute (```topic``` should have preference over the semantic-sugaring of ```targetId```)? I personally would rename ```useFormStyle``` to ```useState``` to signal that visibility rules are treated statefully (and only then ```rulesMethod``` will actually be considered). If not set, ```useState``` should default to false in CoreWidgetProcessing and true for BaseFormControl to keep backwards compatibility (though I would probably patch it to default to true even in CoreWidgetProcessing). If ```useState``` is true, the rules engine would be used, otherwise the good-ol' objectUtils.

**Comment by draperd** [comment]
Ok... thanks for the feedback - I'm at an off-site meeting today so won't be able to make any code updates today, but I'll try and roll in as many of your suggestions as possible - I'll set up a PR for easier review and get your input before requesting a merge.

I agree about not having `topic` and `targetId` either or approach - in fact I found from some early experimentation that it wasn't easily workable from the approach I was taking anyway. 

I'm happy to go with `useState` over `useFormStyle`.

I might take the approach I usually take which is to initially implement the minimum required to satisfy the problem description but ensure that it is open to further iteration.

**Comment by draperd** [comment]
I've opened a PR to make commenting and review of the WIP easier. I've still got to write the new tests but I've changed the attribute name and am supporting both `topic` and `targetId` for non-form control widgets. The code needs tidying generally and the comments need to be improving - but I just wanted to make it simple to follow what I'm up to.

**Comment by draperd** [comment]
I've pushed some more changes and tests. I think we've now got full support for `topic` and `targetId` within the `useState` mode. I think the unit tests cover the majority of scenarios. I'll ask @MartinDoyleUK  to review the code, but hold off on merging to give @AFaust the opportunity to make any further comments/suggestions.

**Comment by draperd** [comment]
The fix for this is in the 1.0.69 release which is now available.
